### ISSUE TITLE
Cannot overfit a single example

### ISSUE THREAD

**Issue by ikvision** [awslabs/gluon-ts#1251]
## Description
Given the introduction example for forecasting I am trying to see if I can get the loss to be negative, but can't even after 1000 epochs - the loss is flat after 50 iterations.
This is a sanity test the system is able to memorize a single example. Without passing this sanity check it is difficult to know if there is a bug in the optimizer

## To Reproduce
```python
import pandas as pd
from gluonts.dataset.common import ListDataset
from gluonts.model.simple_feedforward import SimpleFeedForwardEstimator
from gluonts.mx.trainer import Trainer

url = "https://raw.githubusercontent.com/numenta/NAB/master/data/realTweets/Twitter_volume_AMZN.csv"
df = pd.read_csv(url, header=0, index_col=0)
training_data = ListDataset(
    [{"start": df.index[0], "target": df.value[:"2015-04-05 00:00:00"]}],
    freq = "5min",
)

prediction_length = 1
epochs= 1000
estimator = SimpleFeedForwardEstimator(freq="5min", prediction_length=prediction_length,num_parallel_samples=10,
                                       trainer=Trainer(epochs=epochs,learning_rate=1e-4,minimum_learning_rate=1e-8,
                                                       patience=250,weight_decay=0),
                                       sampling=True,num_hidden_dimensions=[5000],)
predictor = estimator.train(training_data=training_data)

train_forecast_it, train_it = make_evaluation_predictions(
    dataset=training_data,  # test dataset
    predictor=predictor,  # predictor
    num_samples=100,  # number of sample paths we want for evaluation
)
tain_predictor_output_ts_list = list(train_forecast_it)
train_target_gt_ts_list = list(train_it)

for tain_predictor_output_ts, train_target_gt_ts in zip(tain_predictor_output_ts_list,train_target_gt_ts_list):
    train_target_gt_ts[-60:].plot(linewidth=2)
    tain_predictor_output_ts.plot(color='g', prediction_intervals=[50.0, 90.0])
    plt.grid(which='both')
    plt.show()

```

## Error message or code output

```
 epoch=1000/1000, avg_epoch_loss=4.24]
```

## Environment
- Operating system: mac
- Python version: 3.7
- GluonTS version:
- MXNet version:

(Add as much information about your environment as possible, e.g. dependencies versions.)

**Comment by mbohlkeschneider** [comment]
Hi @ikvision,

Thank you for opening this issue. 

Please note that `SimpleFeedForward` will learn the (conditional) parameters of a Student-t output and therefore use the corresponding log-likelihood (https://github.com/awslabs/gluon-ts/blob/9869c68243a45052549c6bf391787d56ef52cac3/src/gluonts/mx/distribution/student_t.py#L83). 

If you assume that you would predict `mu` perfectly, than the student t log-prob would reduce to the z factor, which will be positive if the `sigma` value will be very small (which might be the case for a single example). 

You can check for yourself by running this code: 

```
import numpy as np
import mxnet as mx
def z(nu, sigma): 
    nup1_half = (nu + 1.0) / 2.0
    Z = (
        mx.nd.gammaln(nup1_half)
        - mx.nd.gammaln(nu / 2.0)
        - 0.5 * mx.nd.log(math.pi * nu)
        - mx.nd.log(sigma)
    )
    return Z
nu = np.linspace(0.001, 10, num=100)
sigma = np.linspace(0.001, 0.01, num=100)
all_z = []
for n in nu:
    for s in sigma: 
        all_z.append(student_t_z(mx.nd.array([n]), mx.nd.array([s])).asnumpy()[0])
print(all_z)
```

**Comment by ikvision** [comment]
@mbohlkeschneider 
Thank you for you very quick reply, I didn't find the function `student_t_z ` so I couldn't run the example code you provided.
Can you help me fix my example so the loss get to negative? The optimizer should be able to overfit a single example - this is how generally people debug their input pipelines
Overfit a Single Batch is a [common practice](https://towardsdatascience.com/full-stack-deep-learning-steps-and-tools-a21eda6227b1)

**Comment by mbohlkeschneider** [comment]
Sorry, my bad, the function was included in the snipped but I forgot to change the naming. 

```
import numpy as np
import mxnet as mx
def z(nu, sigma): 
    nup1_half = (nu + 1.0) / 2.0
    Z = (
        mx.nd.gammaln(nup1_half)
        - mx.nd.gammaln(nu / 2.0)
        - 0.5 * mx.nd.log(math.pi * nu)
        - mx.nd.log(sigma)
    )
    return Z
nu = np.linspace(0.001, 10, num=100)
sigma = np.linspace(0.001, 0.01, num=100)
all_z = []
for n in nu:
    for s in sigma: 
        all_z.append(z(mx.nd.array([n]), mx.nd.array([s])).asnumpy()[0])
print(all_z)
```

**Comment by mbohlkeschneider** [comment]
Hi @ikvision,

I don't see why a negative loss would indicate overfitting in this case. Can you elaborate on that?

**Comment by ikvision** [comment]
@mbohlkeschneider  How would you be sure you overfitted the single sample?
        `loss = (-distr.log_prob(target / scale) + torch.log(scale)).mean(dim=1)`
I thought `target==1` means `log_prob(1)==0` and therefore the smaller is the loss - the closer the prediction is to the ground truth

**Comment by lostella** [comment]
@ikvision could you update the snippet in the original post to include the code used to get the forecast & produce the plot?

**Comment by ikvision** [comment]
@lostella I have updated the code in the original post

**Comment by lostella** [comment]
![image](https://user-images.githubusercontent.com/433963/104027500-c3c8f180-51c7-11eb-8d7e-b76698a761af.png)

**Comment by lostella** [comment]
Adapted snippet:

```python
import pandas as pd
import matplotlib.pyplot as plt

from gluonts.dataset.common import ListDataset
from gluonts.evaluation.backtest import make_evaluation_predictions
from gluonts.model.deepar import DeepAREstimator
from gluonts.mx.trainer import Trainer

url = "https://raw.githubusercontent.com/numenta/NAB/master/data/realTweets/Twitter_volume_AMZN.csv"
df = pd.read_csv(url, header=0, index_col=0)
# training_target = list(df[:"2015-04-05 00:00:00"].value)
training_target = list(df[:100].value)

training_data = ListDataset(
    [{"start": df.index[0], "target": training_target}],
    freq="5min",
)

print(f"train ts length = {len(training_target)}")

prediction_length = 6
epochs = 100
batch_size = 32
num_batches_per_epoch = 50

print(f"samples trained on = {epochs * num_batches_per_epoch * batch_size}")

estimator = DeepAREstimator(
    freq="5min",
    prediction_length=prediction_length,
    batch_size=batch_size,
    trainer=Trainer(
        epochs=epochs,
        num_batches_per_epoch=num_batches_per_epoch,
        learning_rate=1e-3,
    ),
)

predictor = estimator.train(training_data=training_data)

train_forecast_it, train_it = make_evaluation_predictions(
    dataset=training_data,  # test dataset
    predictor=predictor,  # predictor
    num_samples=100,  # number of sample paths we want for evaluation
)
train_forecasts = list(train_forecast_it)
train_series = list(train_it)

for f, s in zip(
    train_forecasts, train_series
):
    s[-60:].plot(linewidth=2)
    f.plot(color="g", prediction_intervals=[50.0, 90.0])
    plt.grid(which="both")
    plt.show()
```

**Comment by lostella** [comment]
Significant changes from the original snippet are as follows:
* Using `DeepAREstimator` in place of `SimpleFeedForwardEstimator` (the latter being more intended for illustration purposes)
* Reducing the length of the time series to force data loading to sample more often the sample used for test. This is key: at training time, many of the models load batches composed of samples from all over the time series: if the time series is very long, the model will see quite a few other things before it gets to see the test sample again. This effectively acts as a sort of regularization.

**Comment by ikvision** [comment]
Wow! Thank you so much for this detailed example, it is a leap forward.
Is there another way do decrease the samples sliced randomly from all over the time series (kind of augmentation)? 
I want to reduce this regularization without reducing the length of the time series

**Comment by lostella** [comment]
@ikvision not out of the box (ie, not by changing some option on the estimator) at least for now. In general, you can modify the InstanceSplitter transformation to use a different sampling strategy, and sample much more frequently more recent windows of data. But this require changing the estimator code currently.

**Comment by ikvision** [comment]
@lostella  I think we can close this bug. I those those augmentation are far too aggressive for my use case.
I want to change the InstanceSplitter to cancel all augmentation, just to be sure I am touching the right code - are the augmentation done in?
Do you have a documentation explaining this sampling strategy? it looks unfamiliar to me.
https://github.com/awslabs/gluon-ts/blob/7187e25eda1380d8665f83f6d4c15ec78b8c745c/src/gluonts/model/tft/_transform.py#L93-L183

**Comment by lostella** [comment]
@ikvision sampling strategies are defined in https://github.com/awslabs/gluon-ts/blob/9869c68243a45052549c6bf391787d56ef52cac3/src/gluonts/transform/sampler.py, the `InstanceSplitter` transformation is configured to use one of these when iterating through a set of time series.

For example, the one used by `DeepAREstimator` is https://github.com/awslabs/gluon-ts/blob/9869c68243a45052549c6bf391787d56ef52cac3/src/gluonts/transform/sampler.py#L85

**Comment by ikvision** [comment]
@lostella  It looks pretty rich and a core concept of TsGluon. Why augmentation is such a big part of this framework, is it so significant for improved results?
In computer vision augmentation gives minor boost in performance( 1%-2%) on most standard dataset, is it different for time series?
